fix(auth): Treat authData[provider]=null as unlink; skip provider validation for unlink; prevent re-validation of unchanged providers when linking new auth providers

SNtGog opened the pull request:
#### Pull Request

* Report security issues [[[confidentially](https://github.com/parse-community/parse-server/security/policy)](https://github.com/parse-community/parse-server/security/policy)]([https://github.com/parse-community/parse-server/security/policy](https://github.com/parse-community/parse-server/security/policy)).
* Any contribution is under this [[[license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE)](https://github.com/parse-community/parse-server/blob/alpha/LICENSE)]([https://github.com/parse-community/parse-server/blob/alpha/LICENSE](https://github.com/parse-community/parse-server/blob/alpha/LICENSE)).
* Link this pull request to an [[[issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue)](https://github.com/parse-community/parse-server/issues?q=is%3Aissue)]([https://github.com/parse-community/parse-server/issues?q=is%3Aissue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue)).

#### Issue

Closes: parse-community/parse-server#9855, parse-community/parse-server#9999

---

## Approach

### Summary

Fix incorrect authData validation during `_User` updates that breaks valid multi-provider flows.

* Before calling `Auth.findUsersWithAuthData`, build a filtered object that **excludes providers with `null` or `undefined` values**.
* Apply unlink markers (`authData[provider] = null`) as part of the update **without invoking the provider adapter**.
* Ensure adapter validation runs **only for providers that actually change**; login behavior remains unchanged.

## Behavior & Compatibility

* **Unlinking remains explicit** via `authData[provider] = null`.
* Unchanged providers are no longer revalidated on update.
* Changed providers follow existing validation and conflict checks.
* No changes to session token handling, email verification flow, or public APIs.
* No behavioral changes to login flows.

## Security & Performance

* Prevents unintended adapter calls (e.g. OAuth code → token exchange) for unchanged or unlinked providers.
* Reduces unnecessary external validations in a sensitive auth path without weakening security for real auth changes.

## Tests

* (Added/updated) scenarios:

  * **Multi-provider updates:** add/update one provider while others remain unchanged.
  * **Unlink operations:** `{ provider: null }` unlinks without adapter validation.
  * **Sequential flows:** add → unlink → add.
* Existing tests remain unchanged.

## Release Notes (proposed)

**fix(auth):** Correct authData validation for multi-provider updates. Providers are validated only when modified; unlink (`provider = null`) no longer triggers adapter validation. Login behavior unchanged.

## Tasks

* [x] Add tests
* [ ] Add changes to documentation (guides, repository pages, code comments)
* [ ] Add security checks

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **Bug Fixes**
  * Strengthened provider name validation to reject invalid, reserved, or non-alphanumeric keys.
  * Fixed user account updates via session token authentication.
  * Optimized authentication validation to only check changed providers, reducing unnecessary API calls.

* **New Features**
  * Improved multi-provider account linking and unlinking workflows.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SNtGog committed:
Optimize authData handling logic and integrate delta-based updates

Refactor authData processing for improved clarity, efficiency, and correctness. Introduce `diffAuthData` and `subsetEqual` to handle partial updates, unlinking, and change detection intelligently. Update `RestWrite` logic to utilize refined authData delta behavior, preventing redundant operations. Add comprehensive test coverage for edge cases and optimizations. Refactor existing password policy validations for readability.

parseplatformorg commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/acinader/pr-checks/30e0badb-503a-4c0a-9ee2-e761656a3851?refs=714d80f8006db7634449943566099ae632dfbf08&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

Moumouls commented on src/Auth.js:
```diff
@@ -601,6 +606,70 @@ const handleAuthDataValidation = async (authData, req, foundUser) => {
   return acc;
 };
 
+const subsetEqual = (prev, next) => {
```
issue: diff and equality system should not be coded from scratch and if not done correctly could be an attack vector, popular libs are native implementation should be used

SNtGog replied:
Thanks for the feedback on diff/equality — agreed. I overreached trying to fix too much at once. I’ve reduced this PR to minimal changes only:

Before findUsersWithAuthData, I filter out providers where authData[provider] is null (or undefined), so lookup/validation runs only on non-null providers.

Unlink via authData[provider] = null is applied without invoking the adapter.

If we revisit partial update semantics later, I’ll rely on native or well-vetted utilities and propose that in a separate PR.

Moumouls replied:
Ohh thanks @SNtGog diffs looks much better 🚀

Moumouls reviewed (commented):
note: not in favor of this kind of refactor on a sensible app aera, the "null" issue to unlink a provider should be performed with minimal changes if possible

SNtGog committed:
Revert "Optimize authData handling logic and integrate delta-based updates"

This reverts commit 7669bd9295dae7589d37f90f6de952612bc4df78.

SNtGog committed:
Fix authData handling to exclude unlinked providers and add tests for edge cases

SNtGog committed:
remove redundant tests

Moumouls commented on spec/RestWrite.handleAuthData.spec.js:
```diff
@@ -0,0 +1,67 @@
+    await user.save(null, { sessionToken });
+
+    const updatedUser = await new Parse.Query(Parse.User).get(user.id, { useMasterKey: true });
+    const finalAuthData = updatedUser.get('authData') || {};
```
issue: You should not fall back to "{}". In this test, you expect to receive authData. Otherwise, if authData becomes protected in the future, the test could produce false positives.

Additionally, you need to set another authData in this test to ensure that:
- gpgames is correctly removed without deleting all authData.
- You can then expect that the other authData entries remain unchanged and that the null operation was correctly performed.

Moumouls commented on src/RestWrite.js:
```diff
@@ -538,7 +538,15 @@ RestWrite.prototype.ensureUniqueAuthDataId = async function () {
 RestWrite.prototype.handleAuthData = async function (authData) {
-  const r = await Auth.findUsersWithAuthData(this.config, authData, true);
+  const withoutUnlinked = {};
+  for (const provider of Object.keys(authData)) {
+    if (authData[provider] === null || authData[provider] === undefined) {
+      continue;
+    }
+    withoutUnlinked[provider] = authData[provider];
+  }
```
nitpick: using dedicated tools and functional prog to perform ops on objects, may be you will learn something

const authDataWithoutNullish = Object.fromEntries(Object.entries(authData).filter([_, data] => data ?? false))

When you want to perform filtering on objects, combining Object.fromEntries + Object.entries + .filter() works well (in case of functional programming)

Moumouls replied:
Object.fromEntries combined with Object.entries is underrated: https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries

Moumouls commented:
@SNtGog if you want to fix last details, and we should be good to go here :)

SNtGog committed:
Merge remote-tracking branch 'main/alpha' into refactor-restwrite-authdata-handling

SNtGog committed:
Merge branch 'alpha' into refactor-restwrite-authdata-handling

SNtGog committed:
Refactor authData handling: streamline validation, enhance security, and improve efficiency

SNtGog commented:
Hi @Moumouls,

I’m back — I’ve refactored the changes accordingly and added specs. Please take another look when you have time. Thanks!

Here is some additional context on why the change ended up looking more complex than a simple conditional fix.

The root issue is not a single edge case, but a fundamental mismatch between how authData is currently handled and how it is actually used. The RestWrite class processes multiple distinct semantics (login, linking/updating providers, and unlinking providers) through the same code path, with semantics determined by the presence or absence of query.objectId. These semantics have different and sometimes conflicting validation requirements, yet they share the same validation logic.

In the current implementation, Parse Server does not reliably distinguish between:

providers that are actually being modified,

providers that are unchanged but still present on the user,

providers that are explicitly unlinked via provider: null.

As a result, adapter validation is triggered in cases where it is semantically incorrect (e.g. unchanged providers or unlink operations). This is not just an optimization concern — it breaks valid multi-provider flows entirely for code-based adapters, as demonstrated by the added tests. Even without unlinking (provider: null), adding a second provider while keeping an existing one unchanged can fail because the unchanged provider is revalidated and requires a code that is not part of the update.

Because of this, a “linear” solution (e.g. always validating authData or conditionally skipping validation based on a single flag) is not sufficient. Without computing a delta between stored and incoming authData, Parse Server has no reliable way to determine which providers actually require validation. The refactor introduces this distinction explicitly (changed, unlink, unchanged) to restore correct semantics and make the behavior auditable.

The goal here is not to change behavior arbitrarily, but to make currently broken and untestable scenarios work correctly while preserving existing login behavior. Existing tests remain unchanged; new tests cover valid multi-provider cases that fail without this logic.

SNtGog committed:
Fix incorrect method call in Auth.js (use `master` instead of `Auth.master`)

SNtGog committed:
Remove unused `Auth` import from `rest.js`.

SNtGog committed:
Refactor authData validation tests and helpers: introduce dynamic token resolvers, centralize mock logic, and streamline user/session creation

Moumouls commented:
thanks @SNtGog i'll take a look to the PR again :)

mtrezza commented:
What's the state of this PR?

mtrezza committed:
Merge branch 'alpha' into refactor-restwrite-authdata-handling

mtrezza commented:
I provided a fix with a simpler approach. @SNtGog could you check out https://github.com/parse-community/parse-server/pull/10045 and see whether it works for you?

SNtGog commented:
> I provided a fix with a simpler approach. @SNtGog could you check out #10045 and see whether it works for you?

Hello!

Unfortunately not =((

```describe('RestWrite.handleAuthData', () => {
  const MOCK_USER_ID = 'mockUserId';
  const MOCK_ACCESS_TOKEN = 'mockAccessToken123';

  const createMockUser = () => ({
    id: MOCK_USER_ID,
    code: 'C1',
  });

  const mockGooglePlayGamesAPI = () => {
    mockFetch([
      {
        url: 'https://oauth2.googleapis.com/token',
        method: 'POST',
        response: {
          ok: true,
          json: () => Promise.resolve({ access_token: MOCK_ACCESS_TOKEN }),
        },
      },
      {
        url: `https://www.googleapis.com/games/v1/players/${MOCK_USER_ID}`,
        method: 'GET',
        response: {
          ok: true,
          json: () => Promise.resolve({ playerId: MOCK_USER_ID }),
        },
      },
    ]);
  };

  const setupAuthConfig = () => {
    return reconfigureServer({
      auth: {
        gpgames: {
          clientId: 'validClientId',
          clientSecret: 'validClientSecret',
        },
        instagram: {
          clientId: 'validClientId',
          clientSecret: 'validClientSecret',
          redirectUri: 'https://example.com/callback',
        },
      },
    });
  };

  beforeEach(async () => {
    await setupAuthConfig();
  });

  it('should unlink provider via null', async () => {
    mockGooglePlayGamesAPI();

    const authData = createMockUser();
    const user = await Parse.User.logInWith('gpgames', { authData });
    const sessionToken = user.getSessionToken();

    await user.fetch({ sessionToken });
    const currentAuthData = user.get('authData');
    expect(currentAuthData).toBeDefined();

    mockFetch([
      {
        url: 'https://api.instagram.com/oauth/access_token',
        method: 'POST',
        response: {
          ok: true,
          json: () => Promise.resolve({ access_token: 'ig_token' }),
        },
      },
      {
        url: 'https://graph.instagram.com/me?fields=id&access_token=ig_token',
        method: 'GET',
        response: {
          ok: true,
          json: () => Promise.resolve({ id: 'I1' }),
        },
      },
    ]);

    user.set('authData', {
      ...currentAuthData,
      instagram: { id: 'I1', code: 'IC1' },
    });
    await user.save(null, { sessionToken });

    await user.fetch({ sessionToken });
    const authDataWithInstagram = user.get('authData');
    expect(authDataWithInstagram).toBeDefined();
    expect(authDataWithInstagram.gpgames).toBeDefined();
    expect(authDataWithInstagram.instagram).toBeDefined();

    user.set('authData', {
      ...authDataWithInstagram,
      gpgames: null,
    });
    await user.save(null, { sessionToken });

    const updatedUser = await new Parse.Query(Parse.User).get(user.id, { useMasterKey: true });
    const finalAuthData = updatedUser.get('authData');

    expect(finalAuthData).toBeDefined();
    expect(finalAuthData.gpgames).toBeUndefined();
    expect(finalAuthData.instagram).toBeDefined();
    expect(finalAuthData.instagram.id).toBe('I1');
  });
});
```

 Result:

 % lsof -ti:27017 | xargs kill -9 && npm run build && npm test -- **/RestWrite.handleAuthData.spec.js

> parse-server@9.3.0-alpha.3 build
> babel src/ -d lib/ --copy-files --extensions '.ts,.js'

Successfully compiled 195 files with Babel (1504ms).

> parse-server@9.3.0-alpha.3 test
> cross-env MONGODB_VERSION=${MONGODB_VERSION:=8.0.4} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} mongodb-runner exec -t ${MONGODB_TOPOLOGY} --version ${MONGODB_VERSION} -- --port 27017 -- npm run testonly spec/RestWrite.handleAuthData.spec.js


> parse-server@9.3.0-alpha.3 testonly
> cross-env MONGODB_VERSION=${MONGODB_VERSION:=8.0.4} MONGODB_TOPOLOGY=${MONGODB_TOPOLOGY:=standalone} TESTING=1 jasmine spec/RestWrite.handleAuthData.spec.js

Randomized with seed 09054
Started
Jasmine started
warn: DeprecationWarning: The Parse Server option 'fileUpload.allowedFileUrlDomains' default will change to '[]' in a future version. Set 'fileUpload.allowedFileUrlDomains' to the domains you want to allow, or to '[]' to block all file URLs.
F
  RestWrite.handleAuthData
    ✗ should unlink provider via null
      - Error: gpgames code is required.


Also i would like to provide stacktrace:

beforeFind(), BaseCodeAuthAdapter.js:47
beforeFind(), Auth.js:429
map(), Auth.js:424
findUsersWithAuthData(), RestWrite.js:544
handleAuthData(), RestWrite.js:487
validateAuthData(), RestWrite.js:114
Async call from Promise.then
then(), RestWrite.js:113
execute(), rest.js:315
Async call from Promise.then
then(), rest.js:300
update(), ClassesRouter.js:129
handleUpdate(), ClassesRouter.js:244
handler(), PromiseRouter.js:72
Async call from Promise.then
then(), PromiseRouter.js:71
reduce(), PromiseRouter.js:70
promiseHandler(), PromiseRouter.js:150
next(), middlewares.js:375
next(), middlewares.js:346
Async call from await
handleRateLimit(), middlewares.js:317
next(), middlewares.js:460
next(), middlewares.js:707
next(), middlewares.js:449

mtrezza commented:
@SNtGog are you actually running parse-server based on PR https://github.com/parse-community/parse-server/pull/10045? Did you compile (`npm run watch`)? What is the code line that you see at `BaseCodeAuthAdapter.js:47` from your stack trace?

I've added your test in that PR, and added an additional fix, as it seems to be an additional problem with combined auth providers as in your test. If you test again, please make sure to use that branch with latest commit and compile. If you see an error, please provide the stack trace and the line of code of the top-most stack trace reference.

SNtGog commented:
> @SNtGog are you actually running parse-server based on PR #10045? Did you compile (`npm run watch`)? What is the code line that you see at `BaseCodeAuthAdapter.js:47` from your stack trace?
> I've added your test in that PR, and added an additional fix, as it seems to be an additional problem with combined auth providers as in your test. If you test again, please make sure to use that branch with latest commit and compile. If you see an error, please provide the stack trace and the line of code of the top-most stack trace reference.

I initially tested this on the latest alpha branch, not directly on your PR.

I’ve now checked your PR branch specifically (with a fresh compile), and I can confirm that the issue is resolved there. The problem I was seeing does not reproduce with your changes.

mtrezza commented:
@SNtGog could you try it again with the branch? I've made some changes because tests were failing for Postgres.  Just to verify that it didn't break anything. Thanks!

SNtGog commented:
> @SNtGog could you try it again with the branch? I've made some changes because tests were failing for Postgres. Just to verify that it didn't break anything. Thanks!

Hello! 
Just tested it again with your latest commits — can confirm everything works fine on my side now. Thanks!

parseplatformorg commented:
🎉 This change has been released in version [9.3.0-alpha.4](https://github.com/parse-community/parse-server/releases/tag/9.3.0-alpha.4)

parseplatformorg commented:
🎉 This change has been released in version [9.3.0](https://github.com/parse-community/parse-server/releases/tag/9.3.0)
